### DISCUSSION TITLE
@TestSecurity not working for QuarkusUnitTest in extension

### DISCUSSION THREAD

**Discussion by HerrDerb** [quarkusio/quarkus#38669] [Q&A]
I tried to make a QuarkusUnitTest in a extensions deployment module that uses `@TestSecurity`
This seems not to work. When setting the security identity manually, it works.

```
   @Test
    @ActivateRequestContext
    @TestSecurity(user = "testUser", roles = "testRole")
    @Order(1)
    void whenWithTestSecurityAnnotation_shouldHaveSecurityIdentity() {
        assertEquals(securityIdentity.getPrincipal().getName(),"testUser");   <--- fails here already
        assertTrue( securityIdentity.hasRole("testRole"));
    }

    @Test
    @ActivateRequestContext
    @Order(2)
    void whenProgrammatically_shouldHaveSecurityIdentity() { <--- works
        withUserRole(Set.of("testRole"));

        assertEquals(securityIdentity.getPrincipal().getName(),"testUser"); 
        assertTrue( securityIdentity.hasRole("testRole"));
    }
```

What may be the issue?
I made a reproducer: https://github.com/HerrDerb/extension-issues/tree/test-security-issue

[a comment by an automated account is not shown]

**Comment by sberyozkin** [comment]
@HerrDerb I'm not sure about a test pattern like

```
@Inject
    SecurityIdentity securityIdentity;

    @Test
    @ActivateRequestContext
    @TestSecurity(user = "testUser", roles = "testRole")
    @Order(1)
    void whenWithTestSecurityAnnotation_shouldHaveSecurityIdentity() {
        assertEquals(securityIdentity.getPrincipal().getName(),"testUser");
        assertTrue( securityIdentity.hasRole("testRole"));
    }
```

It should work if you have the identity injected in the secured endpoint, I'm not sure ` @ActivateRequestContext` makes sense at the unit test level.
CC @manovotn @mkouba

**Reply by HerrDerb** [reply]
Without the `@ActivateRequestContext`, you'll get an `ContextNotActiveException`  when running the tests regarding the `SecurityIdentityProxy`

**Reply by mkouba** [reply]
Yes, that's expected - AFAIK we don't activate the request context for a test method automatically.

**Reply by sberyozkin** [reply]
My understanding that there should be something that triggers the request scoped injection, adding `@TestSecurity` makes sure such trigger will work but something has to trigger it, like a request to the secured endpoint

**Reply by HerrDerb** [reply]
You suggest that, to test this properly, I would additionally need to create a test endpoint and make a call?

**Reply by sberyozkin** [reply]
@HerrDerb 

This is how I've been doing it: https://github.com/quarkusio/quarkus/blob/main/integration-tests/oidc-code-flow/src/test/java/io/quarkus/it/keycloak/TestSecurityLazyAuthTest.java
